psbt/psbt.go

```diff
@@ -0,0 +1,676 @@
+// TODO: this is unnecessarily inefficient.
+func (p *PsbtKeyValuePairList) addKVPair(kv *PsbtKeyValuePair) error {
+	for _, kvPair := range p.Items {
+		if areEqualPsbtKeys(kvPair, kv) {
+			return errors.New("Cannot add duplicate key")
+		}
+	}
+
+	p.Items = append(p.Items, kv)
+	p.Sort()
+	return nil
+}
+
+// Created Psbts will have the key-value pairs in any subsection
+// sorted in lexicographical ordering.
+
+// PsbtKeyValueSorter implements sort.Interface.
+type PsbtKeyValueSorter []*PsbtKeyValuePair
+
+func (s PsbtKeyValueSorter) Len() int { return len(s) }
+
+func (s PsbtKeyValueSorter) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+func (s PsbtKeyValueSorter) Less(i, j int) bool {
+	// Must sort by concatenation of key int (as byte) and rest of keydata
+	return bytes.Compare(s[i].getFullKey(), s[j].getFullKey()) < 0
+}
+
+// Sort sorts the PsbtKeyValuePairs in the PsbtKeyValuePairList
+// in lexicographical sort order on the key (=KeyType+keydata).
+func (p *PsbtKeyValuePairList) Sort() {
+	sort.Sort(PsbtKeyValueSorter(p.Items))
+}
+
+// Psbt is a set of 1 + N + M key-value pair lists, 1 global,
+// defining the unsigned transaction structure with N inputs and M outputs.
+// These key-value pairs can contain scripts, signatures,
+// key derivations and other transaction-defining data.
+type Psbt struct {
+	Raw           []byte                  // raw binary serialization of Psbt
+	UnsignedTx    *wire.MsgTx             // Deserialization of unsigned tx
+	GlobalKVPairs *PsbtKeyValuePairList   // One global KVPair set
+	InKVPairs     *[]PsbtKeyValuePairList // Array of KVPair sets, one per input
+	OutKVPairs    *[]PsbtKeyValuePairList // Array of KVPair sets, one per output
+}
+
+// PsbtKeyTypeAndDataFromKey takes a raw byte slice and extracts the KeyType
+// and key data from that slice, checks that the data conforms
+// to the BIP spec, depending on the key type, and section, and returns
+// the KeyType and key data or an error.
+func PsbtKeyTypeAndDataFromKey(keydata []byte, section string) (PsbtKeyType,
+	[]byte, error) {
+	keyType := uint8(string(keydata)[0])
+	if !validateKeyData(keydata[1:], keyType, section) {
+		return 255, nil, errors.New("Invalid keydata for keytype")
+	}
+	// Note that the second return value will usually be empty,
+	// since most keys contain no more than the key type byte.
+	return keyType, keydata[1:], nil
+}
+
+// validateUnsignedTx returns true if the transaction is unsigned.
+// Note that more basic sanity requirements,
+// such as the presence of inputs and outputs, is implicitly
+// checked in the call to MsgTx.Deserialize()
+func validateUnsignedTX(tx *wire.MsgTx) bool {
+	for _, tin := range tx.TxIn {
+		if len(tin.SignatureScript) != 0 || tin.Witness != nil {
+			return false
+		}
+	}
+	return true
+}
+
+// validatePubkey returns true if the passed data is a valid
+// compressed serialization of a valid secp256k1 pubkey.
+// It is a requirement of BIP174 that only compressed pubkeys
+// are allowed.
```

Comment:
I'm not sure this last point is correct - I can't see this anywhere in the BIP, and the reference implementation allows for uncompressed keys

Reply:
Wow, I've no idea what happened here; somehow I had it in my head that this requirement was specified in the BIP, but there nothing saying that there now I read it again ...

It is a requirement of BIP141, I *guess* I may have confused myself by mixing that up, but weird, in any case. Will fix, thanks.

Reply:
Fixed this (check for any valid pubkey) in abf27ad43f22f7556bf9b0c4e7bf5e7b3cb8ea03

Also validates signatures.